### DISCUSSION TITLE
Realm Json Documentation

### DISCUSSION THREAD

**Discussion by igloo15** [keycloak/keycloak#34890] [Q&A]
I am wondering if there is any documentation on the json schema for a realm file. I would like to use realm import but I want to first create a minimalistic realm.json with the base necessary settings. Specifically removing any settings in the json that are default and don't need to be imported unless they are changed from default.

I feel the rest api definitions is an ok source but there is no information on what the properties and items in the json actually mean. For instance AuthenticationFlows has a property called `builtIn` now what does this property mean? I couldn't find any documentation on what it means. I assume it mean the flow is baked into keycloak, if the value is true. Does that mean that unless I am changing it I can just remove it from my realm.json file?

Ideally I would like to start with a blank json file and just add the base minimum changes to get a realm up and running. As far as I can tell though this is extremely difficult. Even the most slim realm.json files are 2000+ lines long.

**Comment by dennisvang** [comment]
I haven't been able to find any clear documentation either, but there are some useful examples in the `keycloak-quickstarts` repo, for example:

- https://github.com/keycloak/keycloak-quickstarts/blob/latest/nodejs/resource-server/config/realm-import.json
- https://github.com/keycloak/keycloak-quickstarts/blob/latest/spring/rest-authz-resource-server/config/realm-import.json

**Comment by driv** [comment]
I've not tried with json, but using the CRD [KeycloakRealmImport](https://www.keycloak.org/operator/realm-import) I was able to create a realm by defining just the name.
```
apiVersion: k8s.keycloak.org/v2alpha1
kind: KeycloakRealmImport
metadata:
  name: test1
spec:
  keycloakCRName: example-ck
  realm:
    realm: test1
 ```

**Comment by ahrens-lgln** [comment]
Currently the best entrypoint seems to be that one:

https://github.com/keycloak/keycloak/blob/main/core/src/main/java/org/keycloak/representations/idm/PartialImportRepresentation.java

It at least shows how the structure may look like, when following the models you are interested in. It tells very little about the fields itself but is better than nothing :)

**Comment by esiha** [comment]
I was wondering the same... and found this: https://github.com/jirutka/keycloak-json-schema
Not official, but close enough for me :)
